File: app/telegram.py
import hmac
import hashlib
import time
import json
from urllib.parse import parse_qsl

def validate_init_data(init_data_raw: str, bot_token: str, max_age_seconds: int = 86400) -> dict:
    pairs = dict(parse_qsl(init_data_raw, keep_blank_values=True))
    hash_value = pairs.get("hash")
    if not hash_value:
        raise ValueError("initData: missing hash")

    pairs.pop("hash", None)
    data_check_string = "\n".join(f"{k}={pairs[k]}" for k in sorted(pairs.keys()))

    secret_key = hmac.new(b"WebAppData", bot_token.encode("utf-8"), hashlib.sha256).digest()
    computed = hmac.new(secret_key, data_check_string.encode("utf-8"), hashlib.sha256).hexdigest()

    if computed != hash_value:
        raise ValueError("initData: bad hash")

    auth_date = int(pairs.get("auth_date", "0"))
    if not auth_date:
        raise ValueError("initData: bad auth_date")
    now = int(time.time())
    if now - auth_date > max_age_seconds:
        raise ValueError("initData: expired")

    user_raw = pairs.get("user")
    user = json.loads(user_raw) if user_raw else None
    return {"user": user, "auth_date": auth_date, "query_id": pairs.get("query_id")}

File: apps/api/src/lib/telegram.ts
import crypto from "crypto";

export type TelegramWebAppUser = {
  id: number;
  first_name?: string;
  last_name?: string;
  username?: string;
  photo_url?: string;
};

export type TelegramInitData = {
  query_id?: string;
  user?: TelegramWebAppUser;
  auth_date: number;
};

function hmacSha256(key: Buffer | string, data: string) {
  return crypto.createHmac("sha256", key).update(data).digest();
}

export function validateInitData(initDataRaw: string, botToken: string, maxAgeSeconds = 86400): TelegramInitData {
  const params = new URLSearchParams(initDataRaw);
  const hash = params.get("hash");
  if (!hash) throw new Error("initData: missing hash");
  params.delete("hash");

  params.sort();
  const pairs: string[] = [];
  for (const [k, v] of params.entries()) pairs.push(`${k}=${v}`);
  const dataCheckString = pairs.join("\n");

  const secretKey = hmacSha256("WebAppData", botToken);
  const computedHash = crypto.createHmac("sha256", secretKey).update(dataCheckString).digest("hex");
  if (computedHash !== hash) throw new Error("initData: bad hash");

  const authDate = Number(params.get("auth_date"));
  if (!Number.isFinite(authDate)) throw new Error("initData: bad auth_date");
  const now = Math.floor(Date.now() / 1000);
  if (now - authDate > maxAgeSeconds) throw new Error("initData: expired");

  const userStr = params.get("user");
  const user = userStr ? (JSON.parse(userStr) as TelegramWebAppUser) : undefined;

  return { query_id: params.get("query_id") ?? undefined, user, auth_date: authDate };
}
